### PR TITLE
Add triangular indices functions

### PR DESCRIPTION
**Pull request by EPronovost**
Add `tril_indices`, `tril_indices_from`, `triu_indices`, and `triu_indices_from`

Mirrors the implementation from numpy

### CODE REVIEW

**Comment by EPronovost** [comment]
Can check off these methods from https://github.com/numba/numba/issues/4074

**Comment by esc** [comment]
@EPronovost thanks for implementing and submitting this. I have queued it for an initial review.

**Comment by esc** [comment]
FYI: CI errors are probably unrelated.

File: numba/targets/arraymath.py [hunk 1]
```diff
@@ -1413,6 +1413,33 @@ def np_tril_impl_multi(m, k=0):
     else:
         return np_tril_impl_multi
 
+@overload(np.tril_indices)
+def np_tril_indices(n, k=0, m=None):
+
+    # we require integer arguments, unlike numpy
```
Comment (esc) [inline numba/targets/arraymath.py]: You might consider putting all the common type-checking for this and the one below (`triu_indices`)  in a re-usable function.

File: numba/tests/test_np_functions.py [hunk 2]
```diff
@@ -124,14 +124,41 @@ def tril_m(m):
...
 
+def tril_indices_n(n):
+    return np.tril_indices(n)
+
+def tril_indices_n_k(n, k=0):
+    return np.tril_indices(n, k)
+
+def tril_indices_n_k_m(n, k=0, m=None):
+    return np.tril_indices(n, k, m)
+
```
Comment (esc) [inline numba/tests/test_np_functions.py]: Maybe you ought to add an additional function here for `np.tril_indices(n, m=None)` that exercises the variant where `k` is omitted?

File: numba/tests/test_np_functions.py [hunk 3]
```diff
@@ -124,14 +124,41 @@ def tril_m(m):
...
 
+def triu_indices_n(n):
+    return np.triu_indices(n)
+
+def triu_indices_n_k(n, k=0):
+    return np.triu_indices(n, k)
+
+def triu_indices_n_k_m(n, k=0, m=None):
+    return np.triu_indices(n, k, m)
+
```
Comment (esc) [inline numba/tests/test_np_functions.py]: Again, a two argument version (`n` and `m`) might be called for.

File: numba/tests/test_np_functions.py [hunk 4]
```diff
@@ -1066,20 +1093,139 @@ def _triangular_matrix_exceptions(self, pyfunc):
...
+        for n in range(10):
+            for m in range(2 * n):
+                for k in range(-n - 1, n + 2):
+                    expected = pyfunc(n, k, m)
+                    got = cfunc(n, k, m)
+                    self.assertEqual(type(expected), type(got))
+                    self.assertEqual(len(expected), len(got))
+                    for e, g in zip(expected, got):
+                        np.testing.assert_array_equal(e, g)
+
```
Comment (esc) [inline numba/tests/test_np_functions.py]: Once you have the `n`  and `m`  versions above, you'll probably want to add an additional test setup here.

File: numba/targets/arraymath.py [hunk 5]
```diff
@@ -1413,6 +1413,33 @@ def np_tril_impl_multi(m, k=0):
...
+    if not isinstance(n, (int, types.Integer)):
+        raise TypeError('n must be an integer')
+    if not isinstance(k, (int, types.Integer)):
+        raise TypeError('k must be an integer')
+    if not (m is None or isinstance(m, (int, types.Integer))):
+        raise TypeError('m must be an integer')
+
+    def np_tril_indices_impl(n, k=0, m=None):
+        return np.nonzero(np.tri(n, m, k=k))
```
Comment (esc) [inline numba/targets/arraymath.py]: Just a question here: the NumPy docs state that 'By default m is taken equal to n.'. Will this happen in the overload implementation of `np.tri`?

**Comment by esc** [comment]
@EPronovost thanks again for submitting this and welcome to Numba! I have reviewed your submission and it looks really good so far. There are a few things that still need to be addressed and I left some comments inline for your consideration. If anything is unclear or confusing, please don't hesitate to ask.

File: numba/targets/arraymath.py (hunk 1 above)
Comment (EPronovost) [reply numba/targets/arraymath.py]: Done

File: numba/targets/arraymath.py (hunk 5 above)
Comment (EPronovost) [reply numba/targets/arraymath.py]: Yes "By default, M is taken equal to N"
https://docs.scipy.org/doc/numpy/reference/generated/numpy.tri.html

Comment (esc) [reply numba/targets/arraymath.py]: Indeed, this is what the docs say. Do you happen to know where Numba does this?

Comment (EPronovost) [reply numba/targets/arraymath.py]: Yes https://github.com/numba/numba/blob/master/numba/targets/arraymath.py#L1360-L1361

File: numba/tests/test_np_functions.py (hunk 3 above)
Comment (EPronovost) [reply numba/tests/test_np_functions.py]: Done

File: numba/tests/test_np_functions.py (hunk 4 above)
Comment (EPronovost) [reply numba/tests/test_np_functions.py]: Done

File: numba/tests/test_np_functions.py (hunk 2 above)
Comment (EPronovost) [reply numba/tests/test_np_functions.py]: Done

**Comment by esc** [comment]
@EPronovost thanks for making these changes and addressing the comments from the review, I will look at them today. Just a word of warning, I saw that you have force pushed the branch, probably due to a git rebase or similar operation. We have had some bad experiences with GitHub loosing review comments and discussions because it can no longer assign them to the respective source lines so we generally discourage the use of rebase. Instead we generally tend to recommend merging `master` into feature branches to update the code base and fix conflicts.

File: numba/targets/arraymath.py [hunk 6]
```diff
@@ -614,6 +614,12 @@ def check_array(a):
         raise ValueError('zero-size array to reduction operation not possible')
 
 
+@register_jitable
+def _check_is_integer(v, name):
+    if not isinstance(v, (int, types.Integer)):
+        raise TypeError('{} must be an integer'.format(name))
```
Comment (esc) [inline numba/targets/arraymath.py]: We generally tend to use `TypingError` in these cases rather than `TypeError`. This is to indicate that Numba failed during type inference.

File: numba/tests/test_np_functions.py [hunk 7]
```diff
@@ -125,6 +125,30 @@ def tril_m_k(m, k=0):
...
+    return np.tril_indices(n, k)
+
+
+def tril_indices_n_m(n, m=None):
+    return np.tril_indices(n, m=m)
+
+
+def tril_indices_n_k_m(n, k=0, m=None):
+    return np.tril_indices(n, k, m)
+
```
Comment (esc) [inline numba/tests/test_np_functions.py]: There is one last corner case that you should address:

```
def tril_indices_n_k_m(n, k=0, m=None):
    return np.tril_indices(n, k=k, m=m)
```

I.e. supplying the arguments as keyword arguments. I recently discovered that this might lead to slightly different path and should be checked to be safe.

File: numba/tests/test_np_functions.py [hunk 8]
```diff
@@ -1066,20 +1114,142 @@ def _triangular_matrix_exceptions(self, pyfunc):
...
+    def _triangular_indices_exceptions(self, pyfunc):
+        cfunc = jit(nopython=True)(pyfunc)
+
+        # Exceptions leak references
+        self.disable_leak_check()
+
+        with self.assertTypingError() as raises:
+            cfunc(4, k=1.5)
+        assert "k must be an integer" in str(raises.exception)
```
Comment (esc) [inline numba/tests/test_np_functions.py]: If possible, can you add some code to test the exceptions for `m` and `n` too?

Comment (stuartarchibald) [reply numba/tests/test_np_functions.py]: Also, please use `self.assertIn()`, Numba employs `unittest` style testing.

**Comment by esc** [comment]
@EPronovost FYI: we would like to get this PR in to the next RC. I have tended to this PR and pushed some commits to your branch so that it has a chance of making it into the next release.

File: numba/targets/arraymath.py [hunk 9]
```diff
@@ -614,6 +614,12 @@ def check_array(a):
         raise ValueError('zero-size array to reduction operation not possible')
 
 
+@register_jitable
```
Comment (stuartarchibald) [inline numba/targets/arraymath.py]: why is this being compiled? the type is known at typing time?

File: numba/targets/arraymath.py [hunk 10]
```diff
@@ -1414,6 +1418,34 @@ def np_tril_impl_multi(m, k=0):
         return np_tril_impl_multi
 
 
+@overload(np.tril_indices)
+def np_tril_indices(n, k=0, m=None):
+
+    # we require integer arguments, unlike numpy
+    _check_is_integer(n, 'n')
+    _check_is_integer(k, 'k')
+    if m is not None:
```
Comment (stuartarchibald) [inline numba/targets/arraymath.py]: This is a `numba.numpy_support.is_nonelike` case, None literal vs NoneType. Merge in master to get that function from it's new API home.

This ought to work:
```python
from numba import njit
import numpy as np


@njit
def triu_indices_n(n, k=0, m=None):
    return np.triu_indices(n, k, m)

print(triu_indices_n(2))
print(triu_indices_n.py_func(2))
```
something like (but with the import at the top of file) will likely sort it:
```python
    from numba.numpy_support import is_nonelike
    if not is_nonelike(m):
        _check_is_integer(m, 'm')
```
same argument applies to the `tril` case.

File: numba/tests/test_np_functions.py [hunk 11]
```diff
@@ -1064,7 +1112,99 @@ def _triangular_matrix_exceptions(self, pyfunc):
...
+                    self.assertEqual(len(expected), len(got))
+                    for e, g in zip(expected, got):
+                        np.testing.assert_array_equal(e, g)
+
+    def _triangular_indices_from_tests_arr_k(self, pyfunc):
+        cfunc = jit(nopython=True)(pyfunc)
+
+        for dtype in [int, float, bool]:
+            for n in range(10):
+                for m in range(10):
```
Comment (stuartarchibald) [inline numba/tests/test_np_functions.py]: `itertools.product`!?

File: numba/tests/test_np_functions.py [hunk 12]
```diff
@@ -1064,7 +1112,99 @@ def _triangular_matrix_exceptions(self, pyfunc):
...
+                    arr = np.ones((n, m), dtype)
+                    expected = pyfunc(arr)
+                    got = cfunc(arr)
+                    self.assertEqual(type(expected), type(got))
+                    self.assertEqual(len(expected), len(got))
+                    for e, g in zip(expected, got):
+                        np.testing.assert_array_equal(e, g)
```
Comment (stuartarchibald) [inline numba/tests/test_np_functions.py]: I'd assume that the arrays returned would be `int` type so `self.assertPreciseEqual` perhaps could be used?

File: numba/tests/test_np_functions.py [hunk 13]
```diff
@@ -1064,7 +1112,99 @@ def _triangular_matrix_exceptions(self, pyfunc):
...
+                        got = cfunc(arr)
+                        self.assertEqual(type(expected), type(got))
+                        self.assertEqual(len(expected), len(got))
+                        for e, g in zip(expected, got):
+                            np.testing.assert_array_equal(e, g)
+
+    def _triangular_indices_exceptions(self, pyfunc):
+        cfunc = jit(nopython=True)(pyfunc)
+
+        # Exceptions leak references
```
Comment (stuartarchibald) [inline numba/tests/test_np_functions.py]: this should be resolvable at typing time?

File: numba/tests/test_np_functions.py [hunk 14]
```diff
@@ -1064,7 +1112,99 @@ def _triangular_matrix_exceptions(self, pyfunc):
...
+            pyfunc,
+            [[n, k, m] for n in range(10) for k in range(-n - 1, n + 2) for m in range(2 * n)]
+        )
+
+    def _triangular_indices_from_tests_arr(self, pyfunc):
+        cfunc = jit(nopython=True)(pyfunc)
+
+        for dtype in [int, float, bool]:
+            for n in range(10):
+                for m in range(10):
```
Comment (stuartarchibald) [inline numba/tests/test_np_functions.py]: `itertools.product`!?

File: numba/tests/test_np_functions.py [hunk 15]
```diff
@@ -1064,7 +1112,99 @@ def _triangular_matrix_exceptions(self, pyfunc):
...
+                    arr = np.ones((n, m), dtype)
```
Comment (stuartarchibald) [inline numba/tests/test_np_functions.py]: this is just testing matrices that are full? given the alg, perhaps put in some zeros via a slice with step or similar?!

File: numba/targets/arraymath.py (hunk 10 above)
Comment (esc) [reply numba/targets/arraymath.py]: yes, the tests don't actually capture this case..

File: numba/tests/test_np_functions.py (hunk 14 above)
Comment (esc) [reply numba/tests/test_np_functions.py]: sure

File: numba/tests/test_np_functions.py (hunk 11 above)
Comment (esc) [reply numba/tests/test_np_functions.py]: yes

File: numba/tests/test_np_functions.py (hunk 13 above)
Comment (esc) [reply numba/tests/test_np_functions.py]: indeed

File: numba/tests/test_np_functions.py (hunk 15 above)
Comment (esc) [reply numba/tests/test_np_functions.py]: You mean randomly or with structure?

Comment (stuartarchibald) [reply numba/tests/test_np_functions.py]: don't think it matters, whatever is convenient.

Comment (stuartarchibald) [reply numba/tests/test_np_functions.py]: actually, ignore this, am remembering a different function!

**Comment by esc** [comment]
Canceled build, will update with new changes shortly.

**Comment by esc** [comment]
@stuartarchibald can you give it another quick look? I think it's almost ready for 'waiting on CI'.

**Comment by EPronovost** [comment]
It seems the tests are failing for Python 2.7 because `inspect.signature` is not defined.  Have you dealt with this in other places?

**Comment by stuartarchibald** [comment]
@esc please could you take a look at the Py2.7 issue, there might be something to help in `numba.six`. Thanks.
